Initial proposal

jorsang1 opened the pull request:
Let's comment over this PR on what we want to improve first
There is a TO-DO section on the readme.
We can add a comment on what we would like to do here so then we can then decide on who does what
Then the idea is to create PR's from/to this branch to then finish the PR with all the suggestions.
Any other approach is also welcome.

jorsang1 committed:
Initial proposal

ensisnoctis commented:
Therefore, officially now, I would like to reserve validation adjustments :)

robertsynoradzki commented:
Therefore, officially now, I would like to reserve validation adjustments :)

jorsang1 commented on README.md:
```diff
@@ -0,0 +1,88 @@
+#### What about the metrics? 
+AppMetrics doesn't seem to be updated for .NET 6.    
+There is this option from MS [Announcing dotnet monitor in .NET 6](https://devblogs.microsoft.com/dotnet/announcing-dotnet-monitor-in-net-6/)   
+
```
More to-do's.....

- Improve validation
- Fix namespaces by adding company name and product/template name
- Add unit testing
- Add logging. 
     Scoped logging, propose the right places on where and what to log (Request filters, mediator behaviors, etc)
- Behavior pipelines for exception handling
- Use records
- Add more examples for Value objects
- Ubiquitous language. Generate dictionary from comments on the Domain layer
- More explanations on the readme or other docs
- ....

jrcbass replied:
I could work on making entities records and struct records ;) @robertsynoradzki be aware that could be some conflicts with the validation. Anyway, I would let you now before starting, maybe on Sunday 24th or probably on Monday 25th

Commit:
Fix build

Commit:
Style: Add .editorconfig

Commit:
Style: Unify usages of final newlines

Commit:
Move properties common for all projects to Directory.Build.props

Commit:
Style: Use file-scoped namespaces everywhere

Commit:
Style: Remove unnecessary usings

Commit:
Style: Unify usage of empty lines

Commit:
Style: Unify usages of newlines

Commit:
Style: Unify usage of semicolons

Commit:
Remove DefineServices pattern

This pattern introduced a risk of duplicate registerations of the same service, possibly with conflicting implementations or scopes

Commit:
Make IDateTime agnostic to timezones and serialization

Commit:
Style: Miscellaneous cleaning

Commit:
Separate validation from domain objects

Commit:
Adjust AuditableEntity

Commit:
Update TODOs in README

robertsynoradzki commented on src/Infrastructure/DependencyInjection.cs:
```diff
@@ -0,0 +1,32 @@
+        services.AddSingleton<IDateTime, DateTimeService>();
+        services.AddSingleton<IProductRepository, ProductRepository>();
+        services.AddSingleton<IStockClient>(provider 
+            => new StockClient(
```
Placing lambda arrow after a newline causing weird indentation dilemmas for the following lines (lambda arrow and space = 3 characters, so how much should we indent the next line while keeping it divisible by 4)? After experimenting with different ideas I've found that lambda arrow on the end of previous line is the least troublesome. I don't mind to force it, but I would like an explicit decision.

robertsynoradzki commented on src/Domain/Entities/Product/Product.cs:
```diff
@@ -0,0 +1,37 @@
+        {
+            AddErrors(Title.Errors);
+        }
+        if (String.IsNullOrEmpty(Sku))
```
```suggestion
        if (string.IsNullOrEmpty(Sku))
```
?

jorsang1 replied:
Done

robertsynoradzki commented on src/Api/Endpoints/ProductEndpoints.cs:
```diff
@@ -0,0 +1,91 @@
+using Application.Products.Queries.GetProduct;
+using Api.Contracts.Responses;
+using Api.Contracts.Requests;
+
```
I strongly believe that our namespaces are already too granular and that "namespaces = folder paths" convention is a code smell. Look at that `using` list - and we've barely started!

jrcbass replied:
Two thoughts:

1. As this is a template, when using Visual Studio or Rider/ReSharper, once the root namespace is chosen, if automatically renamed, all the namespaces will be updated to its consistent folder structure. So it's better to have those using already pointing to that structure.
2.  Also, to avoiding default warnings (yeah, could be disabled) about not matching namespaces, I'll also stick to default folder structure.

All that said, I see some places where it could be interesting to skip that. And of course, I'm also onto that if enough reasons and value :)

robertsynoradzki replied:
The compromise convention I use is that the namespace _must_ reflect the folder structure, but you are free to skip any amount of sub-namespaces from the end, e.g.
Solution `Collaboration.OpenTemplate`, project `Application`, folder `CQRS\Requests\AddProduct`, file `AddproductCommand.cs` can be in either of these namespaces:
- `Collaboration.OpenTemplate.Application.CQRS.Requests.AddProduct`
- `Collaboration.OpenTemplate.Application.CQRS.Requests`
- `Collaboration.OpenTemplate.Application.CQRS`
- `Collaboration.OpenTemplate.Application`
- `Collaboration.OpenTemplate`
And then I just tend towards the shorter ones (results in less sub-namespaces).

Usually I start with one global namespace, then if it gets too crowded, I create a sub-namespace for the biggest group of files that can benefit from a sub-namespace. Repeat when needed.

robertsynoradzki replied:
Anyways, we can change it anytime, and I don't want it to be blocking this PR, so I'll resolve this discussion.

robertsynoradzki commented on .template.config/template.json:
```diff
@@ -0,0 +1,15 @@
+{
+    "$schema": "http://json.schemastore.org/template",
+    "author": "Jordi Sanchez",
```
```suggestion
    "author": "Jordi Sánchez",
```
? :D Or maybe even:
```suggestion
    "author": "Jordi Sánchez García",
```

robertsynoradzki replied:
Don't take it too seriously, I just like the special characters in your name ;)

jorsang1 replied:
ohh man! What a reviewer are you!!
:-D

robertsynoradzki commented on src/Application/Products/Commands/AddProduct/ProductsMapper.cs:
```diff
@@ -0,0 +1,23 @@
+
+internal static class ProductsMapper
+{
+    public static Domain.Entities.Product.Product MapToEntity(this AddProductCommand dto)
```
Extension method for a mapper could be a bad choice for the long term, because we might start requiring:
- injected services,
- mocking and testing.

robertsynoradzki replied:
It is also a bad choice in a sense that we now have 4 different `ProductMapper.cs` files - confusion might arise as the solution grows.

jorsang1 replied:
What do you think would be the best alternative?

jrcbass replied:
Agree, this should change, I have two ideas:

1. A factory
2. An explicit (or implicit) conversion

robertsynoradzki replied:
On second thought, it's okay to have it like that. I'll not be blocking this PR with that petty comment :)

robertsynoradzki commented on src/Domain/Common/AuditableEntity.cs:
```diff
@@ -0,0 +1,12 @@
+﻿namespace Domain.Common;
+
+public abstract class AuditableEntity : Validable
```
Do we only care about the last change? What about history of changes? Perhaps implementing a system-wide notification system instead of method in each class would be more scalable? A notification could be handled by whoever is interested, e.g. a history tracker, or a property updater (instead of forcing each class to update the properties itself). This is a difficult and deep topic I believe. What are the arguments for implementing these properties in the code to begin with?

jorsang1 replied:
You're right, but I think this approach can be a compromise. It offers some kind of minimun amount of auditability while keeping it simple. 
To get a fully audit system then "event sourcing" could be the perfect solution.
Could be fun to implement an ES version of this template, wouldn't it? :-)

robertsynoradzki reviewed (commented):
Other suggestions from me are in the form of a branch that I'll create a PR for.

robertsynoradzki commented on src/Application/Products/DTOs/ProductDto.cs:
```diff
@@ -0,0 +1,3 @@
+﻿namespace Application.Products.DTOs;
```
There is `DTOs` in one namespace but `Api` in another. It would be nice if we consolidated capitalisation :) If we vote, then I vote for `API` in the namespace :)

jorsang1 replied:
I've gone for using `Dto` for now
We could open a discussion, though...

jrcbass replied:
I think this one is important, not only for the naming, but also when using docker. Beware that there can be some problems because of the renamings when using Docker on Linux vs the Windows hold code. So, the sooner we get an agreement for this one, the better :)

robertsynoradzki commented:
This PR is a bag of my other suggestions (so far):
#3

jorsang1 committed:
Adjust namespaces to be replaced using parameters

jorsang1 committed:
Add MediatR as a global using for the Application layer

jorsang1 committed:
Sort some unsorted usings

jorsang1 committed:
Merge pull request #5 from jorsang1/adjust-namespaces-with-parameters

Adjust namespaces to be replaced using parameters

jorsang1 committed:
Adapt 'Notification pattern' section on readme.md

Adapt explanation of the notification pattern for validations in the readme

jorsang1 committed:
Merge pull request #3 from jorsang1/adjustments-proposal

Adjustments to initial proposal

robertsynoradzki commented on src/Domain/Common/ValueObject.cs:
```diff
@@ -0,0 +1,40 @@
+﻿namespace Domain.Common;
+
+// Learn more: https://docs.microsoft.com/en-us/dotnet/standard/microservices-architecture/microservice-ddd-cqrs-patterns/implement-value-objects
+public abstract class ValueObject : Validable
```
There are also some libraries for `ValueObject`s, but I don't know if I like'em. I would prefer creating them manually and having full control.

jorsang1 replied:
Added to the TODO's
Feel free to take this task :-)

robertsynoradzki commented on src/Application/Products/Commands/UpdateProduct/UpdateProductCommandHandler.cs:
```diff
@@ -0,0 +1,61 @@
+    public async Task<ProductDto?> Handle(UpdateProductCommand request, CancellationToken cancellationToken)
+    {
+        var productToUpdate = request.MapToEntity();
+        UpdateAuditableInformation(productToUpdate);
```
(I was putting my comment on line 25 - GitHub displays 22-25. That's also the case in my other comments. Pity that I cannot select a block I'm commenting nor even multiple lines).

jorsang1 committed:
Some more adjustments

Apply some feedback from pull request

jorsang1 commented on src/Api/DependencyInjection.cs:
```diff
@@ -0,0 +1,33 @@
+
+public static class DependencyInjection
+{
+    private static readonly ProductEndpoints productEndpoints = new ProductEndpoints();
```
Done

jorsang1 commented on src/Application/Products/Commands/UpdateProduct/UpdateProductCommandHandler.cs:
```diff
@@ -0,0 +1,61 @@
+    private void UpdateAuditableInformation(Domain.Entities.Product.Product productToAdd)
+    {
+        productToAdd.LastModified = _dateService.Now;
+        productToAdd.LastModifiedBy = Guid.Empty.ToString();
```
I've left this with Guid for now.
There is a TODO to tackle the ID primitive obsession topic

jorsang1 commented on src/Api/Dockerfile:
```diff
@@ -0,0 +1,22 @@
+#See https://aka.ms/containerfastmode to understand how Visual Studio uses this Dockerfile to build your images for faster debugging.
```
Added to the to-do list
Feel free to take it
:-)

jsellaresi committed:
FluentValidation proposal in Domain Layer

jsellaresi committed:
Update src/Domain/Entities/Product/ValueObjects/ProjectTitleValidator.cs

Accepted option

Co-authored-by: jorsang1 <jorsang1@users.noreply.github.com>

jsellaresi committed:
PR suggestions

jorsang1 committed:
Introduce some fixes and refactor mappers

jsellaresi committed:
Typo

jsellaresi committed:
Constructor injections better readability

jorsang1 committed:
Merge pull request #7 from jorsang1/some-fixes

Some fixes

jorsang1 committed:
Merge pull request #6 from jorsang1/fluent-validation-proposal

Fluent validation proposal

jorsang1 committed:
Add some more adjustments

jsellaresi committed:
Testing architecture proposal

jorsang1 committed:
Add scope logging

jorsang1 committed:
Fix dockerfile and add instructions to run the application with docker.

jsellaresi committed:
Expanding tests to achieve better mutation scores

jorsang1 committed:
Apply feedback from reviewers

jorsang1 committed:
Merge pull request #8 from jorsang1/add-scope-logging

Add scope logging

jsellaresi committed:
PR Clear

jsellaresi committed:
Documentation updated

jorsang1 committed:
Merge pull request #9 from jorsang1/testing-proposal

Testing proposal

jorsang1 committed:
Strong typed entity Ids

Primitive obsession on IDs

jrcbass committed:
Add Dockerfile

jrcbass committed:
Add docker-docker-compose

jorsang1 committed:
Improve readme.md and some ADR decissions.

jsellaresi committed:
Multi project runner with relative paths.

jorsang1 committed:
Merge pull request #10 from jorsang1/typed-entity-ids

Typed entity ids

jorsang1 committed:
Merge pull request #11 from jorsang1/add-docker-compose

Added Docker container orchestation

jorsang1 committed:
Merge pull request #12 from jorsang1/stryker-multiple-project-runner

Multi project runner with relative paths.

jorsang1 committed:
Adapt the readme to docker compose recently added.

jsellaresi committed:
Adding DynamicProxyGenAssembly2 to Directory.Build.props to avoid having 2 HandlerExposeds.

jsellaresi committed:
Boyscouting

jorsang1 committed:
Add infrastructure tests

jorsang1 committed:
Merge pull request #14 from jorsang1/dynamicProxyGenAssembly-tests-visibility

Dynamic proxy gen assembly tests visibility

jrcbass committed:
ddd records

jsellaresi committed:
use Stryker on Infrastructure tests

jrcbass committed:
Product creation, update to domain instead of map.

jsellaresi committed:
Boyscouting with better code organization.

jrcbass committed:
Record struct's to sealed record's

jorsang1 committed:
Improve tests names

jorsang1 committed:
Merge pull request #15 from jorsang1/add-infrastructure-tests

Add infrastructure tests

jorsang1 committed:
Add open telemetry

jorsang1 committed:
Merge pull request #17 from jorsang1/add-open-telemetry

Add open telemetry

jorsang1 committed:
Merge pull request #16 from jorsang1/boyscouting-application-unittests

Boyscouting with better code organization.

jrcbass committed:
Updated domain objects and models

jrcbass committed:
Update some tests

jrcbass committed:
Update queries to sealed record's

jorsang1 committed:
Merge pull request #18 from jorsang1/ddd_records

Ddd records

jorsang1 committed:
Add metrics handling

jrcbass committed:
Not domain records to readonly record struct's

jrcbass committed:
UserId creation from commands

jrcbass committed:
ProductListItemDto for list of Product items

jrcbass committed:
Move Datetime from domain to App - Infrastructure

jrcbass committed:
Update ProductBuilder methods to product ctor

jrcbass committed:
Rename DateTimeService

jorsang1 committed:
Merge pull request #19 from jorsang1/Updates_on_records_and_domain

Updates on records and domain

jsellaresi committed:
Fixing Stryker wrong behavior.
